zebra-network/src/meta_addr/tests/prop.rs

```diff
@@ -188,4 +210,209 @@ proptest! {
+                                          .and_then(|v| v.parse().ok())
+                                          .unwrap_or(DEFAULT_VERBOSE_TEST_PROPTEST_CASES)))]
+
+    /// Make sure that [`MetaAddr`]s do not get retried more than once per
+    /// [`LIVE_PEER_DURATION`], regardless of the [`MetaAddrChange`]s that are
+    /// applied to a single peer's entries in the [`AddressBook`].
+    ///
+    /// This is the complex version of the test, which checks [`MetaAddr`],
+    /// [`CandidateSet`] and [`AddressBook`] together.
+    #[test]
+    fn individual_peer_retry_limit_candidate_set(
+        (addr, changes) in MetaAddrChange::addr_changes_strategy(MAX_ADDR_CHANGE)
+    ) {
+        zebra_test::init();
+
+        // Run the test for this many simulated live peer durations
+        const LIVE_PEER_INTERVALS: u32 = 3;
+        // Run the test for this much simulated time
+        let overall_test_time: Duration = LIVE_PEER_DURATION * LIVE_PEER_INTERVALS;
+        // Advance the clock by this much for every peer change
+        let peer_change_interval: Duration = overall_test_time / MAX_ADDR_CHANGE.try_into().unwrap();
+
+        assert!(
+            u32::try_from(MAX_ADDR_CHANGE).unwrap() >= 3 * LIVE_PEER_INTERVALS,
+            "there are enough changes for good test coverage",
+        );
+
+        let runtime = Runtime::new().expect("Failed to create Tokio runtime");
+        let _guard = runtime.enter();
+
+        // Only put valid addresses in the address book.
+        // This means some tests will start with an empty address book.
+        let addrs = if addr.last_known_info_is_valid_for_outbound() {
+            Some(addr)
+        } else {
+            None
+        };
+
+        let address_book = Arc::new(std::sync::Mutex::new(AddressBook::new_with_addrs(&Config::default(), Span::none(), addrs)));
+        let peer_service = service_fn(|_| async { unreachable!("Service should not be called") });
+        let mut candidate_set = CandidateSet::new(address_book.clone(), peer_service);
+
+        runtime.block_on(async move {
+            tokio::time::pause();
+
+            // The earliest time we can have a valid next attempt for this peer
+            let earliest_next_attempt = Instant::now() + LIVE_PEER_DURATION;
+
+            // The number of attempts for this peer in the last LIVE_PEER_DURATION
+            let mut attempt_count: usize = 0;
+
+            for (i, change) in changes.into_iter().enumerate() {
+                while let Some(candidate_addr) = candidate_set.next().await {
+                    assert_eq!(candidate_addr.addr, addr.addr);
+
+                    attempt_count += 1;
+                    assert!(
+                        attempt_count <= 1,
+                        "candidate: {:?}, change: {}, now: {:?}, earliest next attempt: {:?}, attempts: {}, live peer interval limit: {}, test time limit: {:?}, peer change interval: {:?}, original addr was in address book: {}",
```

Comment:
Maybe a bit nitpicky, but maybe it helps to use `concat!` to split this long line?

Reply:
I always forget you can just split and escape string literals however you like.

Escaped newlines and intents produce exactly the same string:
https://github.com/ZcashFoundation/zebra/blob/5fcbfa8065fcc9c1dd3161767eae8539a9adb1fa/zebra-network/src/meta_addr/tests/prop.rs#L313-L315

I had to go look this up: https://doc.rust-lang.org/rust-by-example/std/str.html#literals-and-escapes

Reply:
Oh yeah, that's definitely better 🎉